### DISCUSSION TITLE
Project Proposal: Parametric Namespaces

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#5326] [Ideas]
# Short Summary
Add parameters to namespaces (that are always fully applied) which is implicitly available throughout the namespace.

# Motivating Example

## The Problem
I want to create a library for good-looking terminal functionality, like prompting yes/no questions or asking the user to choose between options in a list. This library will then handle re-prompts in case the user types something nonsensible. I want this library to take advantage of ANSI terminal functionality (like clearing the terminal between prompts) if possible. This might look something like this:
```scala
namespace SmartTerminal {
    pub def promptYesNo(ansi: Bool, prompt: String): Bool \ IO = ???
    pub def promptOptions(ansi: Bool, options: List[String]): String \ IO = ???
    pub def getInput(ansi: Bool, inputChecker: String -> Bool): String \ IO = ??? 
    pub def output(ansi: Bool, msg: String): Unit \ IO = ???
    // ...
}
```
Okay, works decently, but it's a bit annoying to have `ansi` everywhere (ignore for generality that it could be found through system calls, maybe the user wants to disable it anyway). Now let's look at a use site:
```scala
def main(): Unit \ IO = {
    let ansi = ???; // some terminal capability things
    let happy = SmartTerminal.promptYesNo(ansi, "Are you happy in life?");
    SmartTerminal.output(ansi, if (happy) "I'm glad to hear that" else "oh... moving on");
    // ...
}
```

It is again annoying that we have to keep applying `ansi`. It's even worse when we at the use-site need to carry around this `ansi` value. What could we do? in Scala, we could keep it as a static value, or we could add it as an implicit value/argument. This is not possible in Flix.

## The Solution
If I were forced to handle this in current flix, maybe I would have a single builder function in `SmartTerminal` that receives an `ansi: Bool` argument and would then return a huge record that contains all the functions from above, except that they now have the knowledge of `ansi` from their scope.

I think this idea is fine conceptually but painful in practice. But what if we just did this same thing to namespaces themselves?

```scala
namespace SmartTerminal(ansi: Bool) {
    pub def promptYesNo(prompt: String): Bool \ IO = ???
    pub def promptOptions(options: List[String]): String \ IO = ???
    pub def getInput(inputChecker: String -> Bool): String \ IO = ??? 
    pub def output(msg: String): Unit \ IO = ???
    // ...
}
```
Really nice, right? use site: (Here we are in the future where namespace aliasing exists, if not available, I'd have to import the functions by name)
```scala
def main(): Unit \ IO = {
    let ansi = ???; // some terminal capability things
    use {SmartTerminal(ansi) => ST};
    let happy = ST.promptYesNo("Are you happy in life?");
    ST.output(if (happy) "I'm glad to hear that" else "oh... moving on");
    // ...
}
```
I think this would be a really nice feature to have. The hope is that this version of the code could very quickly be transformed back into the first version, thus making sure that monomorphization etc. still works.

# Other Concerns

## Parametric import across functions
If I, at the use site, want to call a function from main that also needs to use the smart terminal, I'd still have to supply an `ansi` argument. A fix to that would be a separate inner namespace. This means that main finds the value of `ansi` and then calls `Inner(ansi).doWhatever(..)` where all the code that needs ansi knowledge lives, this now turns `ansi` into an implicit thing. This seems like a good enough solution for me, but other more sophisticated solutions could exist.

## Oblivious Functions
What if, in the smart terminal namespace, want to have a function that doesn't care about ANSI? Then it seems annoying to have to call it using `SmartTerminal(true).f(...)` because the boolean value supplied is never used. This could be supported by a modifier like so
```scala
oblivious def f(...): Bool = ???
```
This would mean that the body of `f` cannot use the namespace parameter but can now be referenced, both as `SmartTerminal.f` but also hopefully as `SmartTerminal(whatever).f` to not lose the function in contexts like the one above where i use an aliased instantiated namespace.

## Multiple Namespace Parameters
A namespace cannot ever be partially applied or used as a first-class thing, but multiple parameters could make things like oblivious functions more complicated (maybe you can be oblivious to some parameters? are they then by-name?)

**Comment by paulbutcher** [comment]
I think that the problem you're trying to solve is very similar in motivation to the one I had in mind in this conversation on Gitter?:

<img width="866" alt="Untitled" src="https://user-images.githubusercontent.com/61163/211849998-809afdad-b9c3-4db2-9c7e-92a0d1b5868d.png">

**Comment by paulbutcher** [comment]
@JonathanStarup would first-class memoization solve this use case?

**Comment by JonathanStarup** [comment]
First-class memoization to me sounds like caching outputs based on inputs and I'm not sure this supports that? But it looks like this idea would be very useful for https://github.com/paulbutcher/flix-api/blob/master/src/Namespace.flix. Its nice already to have a usecase :D

**Comment by magnus-madsen** [comment]
I think implicit parameters are the best solution. The second best would be algebraic effects. A third solution could be to simply return a record with `promptYesNo`, `promptOptions`, and so forth fields?

**Comment by magnus-madsen** [comment]
Alternatively, if its only a bool, why not just use two namespaces and switch between them?

**Comment by JonathanStarup** [comment]
> I think implicit parameters are the best solution.

Very possible yea, I was just unsure about the implications of full on implicit parameters. Are implicit parameters in Flix' future? The downside here is also that you still have to write it on all the defs.

> The second best would be algebraic effects

I think this could also be good yes

> A third solution could be to simply return a record with promptYesNo, promptOptions, and so forth fields?

I don't think this is a very nice solution. It would break normal documentation tooling and seems very cumbersome.

> Alternatively, if its only a bool, why not just use two namespaces and switch between them?

Copy the whole namespace? Also doesn't seem very nice and annoying to maintain
